metomi/isodatetime/data.py

```diff
@@ -1217,314 +1407,240 @@ def add_truncated(self, year_of_century=None, year_of_decade=None,
+        new_timepoint._time_zone = self._time_zone._copy()
+        return new_timepoint
+
+    def get_props(self) -> list:
+        """Return the data properties of this TimePoint as a list of tuples."""
+        props = []
+        for attr in self.__slots__:
             value = getattr(self, attr, None)
-            if callable(getattr(value, "copy", None)):
-                value = value.copy()
-            hash_.append((attr, value))
-        return hash_
+            if callable(getattr(value, "_copy", None)):
+                value = value._copy()
+            props.append((attr[1:], value))
+            # Have sliced attr string to remove leading underscore
+        return props
 
-    def __eq__(self, other: "TimePoint") -> bool:
-        if other is None:
-            return False
-        if self.truncated != other.truncated:
-            raise TypeError(
-                "Cannot compare truncated to non-truncated " +
-                "TimePoint: %s, %s" % (self, other))
+    def __hash__(self) -> int:
+        if self._truncated:
+            # TODO: Convert truncated TimePoints to UTC when not buggy
+            return hash(
+                tuple(getattr(self, attr) for attr in self.__slots__))
+        point = self.to_utc()
+        return hash((*point.get_calendar_date(),
+                     *point.get_hour_minute_second()))
+
+    def _cmp(self, other: "TimePoint", op: str) -> bool:
+        """Compare self with other, using the chosen operator.
+
+        Args:
+            op: The comparison operator/method, one of ["eq", "lt", "le",
+                "gt", "ge"].
+        """
+        if not isinstance(other, TimePoint):
+            return NotImplemented
+        if self._truncated != other._truncated:
+            raise ValueError(
+                "Cannot compare truncated to non-truncated "
+                "TimePoint: {0}, {1}".format(self, other))
         if self.get_props() == other.get_props():
-            return True
-        if self.truncated:
-            for attribute in self.DATA_ATTRIBUTES:
-                other_attr = getattr(other, attribute)
+            return True if op in ["eq", "le", "ge"] else False
+        if self._truncated:
+            # TODO: Convert truncated TimePoints to UTC when not buggy
+            for attribute in self.__slots__:
                 self_attr = getattr(self, attribute)
+                other_attr = getattr(other, attribute)
                 if self_attr != other_attr:
-                    return self_attr == other_attr
+                    if op == "eq":
+                        return False
+                    if op == "lt":
+                        return self_attr < other_attr
+                    if op == "le":
+                        return self_attr <= other_attr
+                    if op == "gt":
+                        return self_attr > other_attr
+                    if op == "ge":
+                        return self_attr >= other_attr
```

Comment:
I **think** later it could be possible to use the `operator` module? Something like have a dict of operators (e.g. `operator.lt.__name__ => operator.lt`and etc.  I think?). Then just `op = dict.get(op); return op(self_attr, other_attr)`.